### DISCUSSION TITLE
`nvim-cmp` + `copilot.nvim`, suggestion breaks code when selecting copilot source suggestion

### DISCUSSION THREAD

**Discussion by Nikola-Milovic** [LazyVim/LazyVim#3070] [Q&A]
Hello everyone, I have had this issue for the longest time, not sure what causes it so not sure where to submit an issue/ discussion. Is it `nvim-cmp`, `copilot.nvim` or something else?

Steps to reproduce for me are 
1. Use golang and have some kind of a struct
2. Get autocomplete from copilot 
![cmp suggestion](https://github.com/LazyVim/LazyVim/assets/50072027/7d531b4a-fca3-4c6c-a111-ac96c74fcf75)
3.`<C>+k` to move up and get the suggestion
![broken code](https://github.com/LazyVim/LazyVim/assets/50072027/bcb396d2-da64-4ca9-8cb1-e71ad38fe903)

This will result in the lines getting messed up, its a bit hard to see, but basically two lines get merged for some reason.

Here's my [nvim config](https://github.com/Nikola-Milovic/.dotfiles/tree/master/nvim/.config/nvim)

And my `cmp` config, with super tab + custom hotkeys for moving up and down

```lua
return {
  {
    "L3MON4D3/LuaSnip",
    keys = function()
      return {}
    end,
  },
  {
    "hrsh7th/nvim-cmp",
    ---@param opts cmp.ConfigSchema
    opts = function(_, opts)
      local has_words_before = function()
        unpack = unpack or table.unpack
        local line, col = unpack(vim.api.nvim_win_get_cursor(0))
        return col ~= 0 and vim.api.nvim_buf_get_lines(0, line - 1, line, true)[1]:sub(col, col):match("%s") == nil
      end

      local luasnip = require("luasnip")
      local cmp = require("cmp")

      local cmp_window = require("cmp.config.window")
      opts.window = {
        completion = cmp_window.bordered(),
        documentation = cmp_window.bordered(),
      }

      opts.mapping = vim.tbl_extend("force", opts.mapping, {
        ["<C-j>"] = cmp.mapping.select_next_item({ behavior = cmp.SelectBehavior.Insert }),
        ["<C-k>"] = cmp.mapping.select_prev_item({ behavior = cmp.SelectBehavior.Insert }),
        ["<C-b>"] = cmp.mapping.scroll_docs(-4),
        ["<C-f>"] = cmp.mapping.scroll_docs(4),
        ["<C-Space>"] = cmp.mapping.complete({ reason = cmp.ContextReason.Auto }),
        ["<C-e>"] = cmp.mapping.abort(),
        ["<CR>"] = cmp.mapping.confirm({ select = true }), -- Accept currently selected item. Set `select` to `false` to only confirm explicitly selected items.
        ["<S-CR>"] = cmp.mapping.confirm({
          behavior = cmp.ConfirmBehavior.Replace,
          select = true,
        }),
        ["<Tab>"] = cmp.mapping(function(fallback)
          if cmp.visible() then
            -- You could replace select_next_item() with confirm({ select = true }) to get VS Code autocompletion behavior
            cmp.select_next_item()
          -- You could replace the expand_or_jumpable() calls with expand_or_locally_jumpable()
          -- this way you will only jump inside the snippet region
          elseif luasnip.expand_or_jumpable() then
            luasnip.expand_or_jump()
          elseif has_words_before() then
            cmp.complete()
          else
            fallback()
          end
        end, { "i", "s" }),
        ["<S-Tab>"] = cmp.mapping(function(fallback)
          if cmp.visible() then
            cmp.select_prev_item()
          elseif luasnip.jumpable(-1) then
            luasnip.jump(-1)
          else
            fallback()
          end
        end, { "i", "s" }),
      })
    end,
  },
}
```

After some debugging it also seems that the problem is not in the hotkeys, but it gets messed up whenever I insert the suggestion from copilot.

**Comment by richardvanbergen** [comment]
I don't have an answer but I'm guessing you might have better luck searching/asking in the `copilot.lua` repo. At least in VSCode it seems to know where to replace up to whereas this plugin doesn't. It just get's inserted via cmp.

I think for LazyVim it might be better to somehow leave it out `nvim-cmp` and only trigger it explicitly. It just doesn't feel very refined at the moment.

**Reply by Nikola-Milovic** [reply]
Thanks @richardvanbergen, I had to post it here since this community usually is really helpful and I wasn't 100% whether this is `copilot.lua` specific.
